### PR TITLE
Trigger deprecation notices when inherited class calls parent method but misses adding new arguments

### PR DESCRIPTION
**Pull request by kevinjhappy**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes 
| Fixed tickets | 
| License       | MIT
| Doc PR        |

This Pull Request concern severals components, the purpose here is to notify in dev mode that in a case of inherit class, a function will have a new or severals arguments in Symfony 5.0, therefore not implement it is deprecated since Symfony 4.2

The function is made by these conditions : 
1- ```(func_num_args() < $x)``` where [x] is the number of arguments we will have in Symfony 5.0
  this check allow to verify that the arguments are missing 
2- ```(__CLASS__ !== \get_class($this))```
  this check allow to verify that the name of the class is different than the base class, therefore that we are in the child class
3- ```(__CLASS__ !== (new \ReflectionMethod($this, __FUNCTION__))->getDeclaringClass()->getName())```
  this check allow to verify that the class of the current function is different than the base class, therefore the function has been rewrote into the child class

Code exemple : 
``` 
public function method(/* void $myNewArgument = null */)
{
	if ((func_num_args() < 1) && (__CLASS__ !== \get_class($this)) && (__CLASS__ !== (new \ReflectionMethod($this, __FUNCTION__))->getDeclaringClass()->getName())){
            @trigger_error(sprintf('The "%s()" method will have one `void $myNewArgument = null` argument in version 5.0 and higher.Not defining it is deprecated since Symfony 4.2.', __METHOD__ ), E_USER_DEPRECATED);
    }
    // do something 
}
```

The unit test are made by creating a child Class using for the tested function ```return parent::function()``` and by calling this child class and catching the expected depreciation message

### CODE REVIEW

File: src/Symfony/Bridge/Monolog/Logger.php [hunk 1]
```diff
@@ -28,6 +28,10 @@ class Logger extends BaseLogger implements DebugLoggerInterface, ResetInterface
      */
     public function getLogs(/* Request $request = null */)
     {
+        if ((func_num_args() < 1) && (__CLASS__ !== \get_class($this)) && (__CLASS__ !== (new \ReflectionMethod($this, __FUNCTION__))->getDeclaringClass()->getName())){
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/Logger.php]: please remove all extra brackets (same below)

File: src/Symfony/Bridge/Monolog/Logger.php [hunk 2]
```diff
@@ -28,7 +28,7 @@ class Logger extends BaseLogger implements DebugLoggerInterface, ResetInterface
      */
     public function getLogs(/* Request $request = null */)
     {
-        if ((func_num_args() < 1) && (__CLASS__ !== \get_class($this)) && (__CLASS__ !== (new \ReflectionMethod($this, __FUNCTION__))->getDeclaringClass()->getName())) {
+        if (func_num_args() < 1 && __CLASS__ !== \get_class($this) && __CLASS__ !== (new \ReflectionMethod($this, __FUNCTION__))->getDeclaringClass()->getName()) {
             @trigger_error(sprintf('The "%s()" method will have one `Request $request = null` argument in version 5.0 and higher.Not defining it is deprecated since Symfony 4.1.', __METHOD__), E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/Logger.php]: ...`, not defining it`... (same below

Comment (nicolas-grekas) [reply src/Symfony/Bridge/Monolog/Logger.php]: since Symfony 4.2.

File: UPGRADE-4.2.md [hunk 3]
```diff
@@ -75,6 +75,8 @@ Process
    // alternatively, when a shell wrapper is required
    $process = Process::fromShellCommandline('ls -l');
    ```
+ * Add trigger_error messages with E_USER_DEPRECATED to functions in components Bridge/Monolog, BrowserKit, DomCrawler, Finder and Serializer when class is inherited and
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: there should be one line per affected component, telling roughly the same as what deprecation messages tell (same in UPGRADE-5.0)

File: src/Symfony/Bridge/Monolog/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 -----
 
  * added `ProcessorInterface`: an optional interface to allow autoconfiguration of Monolog processors
+ * [DEPRECATION] add trigger_error in case of inherit class using getLogs and countErrors
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/CHANGELOG.md]: no need for the `[...]` label
also the message should be less technical and more useful to the reader (stating roughly the same as the deprecation message)

File: src/Symfony/Component/BrowserKit/CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+4.2.0
+-----
+
+ * [DEPRECATION] add trigger_error in case of inherit class using submit
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/CHANGELOG.md]: (same as above)

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php [hunk 6]
```diff
@@ -54,10 +54,8 @@ public function __construct(TokenStorageInterface $tokenStorage, EventDispatcher
      *
      * @param string $providerKey The name of the provider/firewall being used for authentication
      */
-    public function authenticateWithToken(TokenInterface $token, Request $request/*, string $providerKey */)
+    public function authenticateWithToken(TokenInterface $token, Request $request, string $providerKey)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: hum actually we should add null as a default value here, so not break BC at least (and thus revert the changes on the test cases)

Comment (kevinjhappy) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: No problem to make this change, I just observed in the project Symfony that all the features using this function implement the providerKey argument and as the class is annotate final I believed it will be OK, only the unit tests was needed to be changed

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: if anyone else calls the function, their code will break. since the method is not internal, that's should remain supported.

File: src/Symfony/Bridge/Monolog/Logger.php [hunk 7]
```diff
@@ -28,6 +28,10 @@ class Logger extends BaseLogger implements DebugLoggerInterface, ResetInterface
      */
     public function getLogs(/* Request $request = null */)
     {
+        if (\func_num_args() < 1 && __CLASS__ !== \get_class($this) && __CLASS__ !== (new \ReflectionMethod($this, __FUNCTION__))->getDeclaringClass()->getName()) {
```
Comment (stof) [inline src/Symfony/Bridge/Monolog/Logger.php]: I would add `&& !$this instanceof \PHPUnit\Framework\MockObject\MockObject && !$this instanceof \Prophecy\Prophecy\ProphecySubjectInterface` to avoid triggering this when mocking the class with PHPUnit or Prophecy (the mock class will get its signature from the actual signature, so the argument won't be added until we add it for real, but that's fine as it will then be added automatically)

Comment (stof) [reply src/Symfony/Bridge/Monolog/Logger.php]: that applies to all places of course

File: src/Symfony/Component/Lock/LockInterface.php [hunk 8]
```diff
@@ -43,7 +43,7 @@ public function acquire($blocking = false);
      * @throws LockConflictedException If the lock is acquired by someone else
      * @throws LockAcquiringException  If the lock can not be refreshed
      */
-    public function refresh(/* $ttl = null */);
+    public function refresh($ttl = null);
```
Comment (stof) [inline src/Symfony/Component/Lock/LockInterface.php]: that change is a BC break. We cannot uncomment it until 5.0

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php [hunk 9]
```diff
@@ -54,10 +54,8 @@ public function __construct(TokenStorageInterface $tokenStorage, EventDispatcher
      *
      * @param string $providerKey The name of the provider/firewall being used for authentication
      */
-    public function authenticateWithToken(TokenInterface $token, Request $request/*, string $providerKey */)
+    public function authenticateWithToken(TokenInterface $token, Request $request, string $providerKey = null)
```
Comment (stof) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: this is also a BC break

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
Most of these deprecations won't be triggered if the child method does not call the parent method, so they might not be the best place to trigger them (sometimes, we might not have a better way though)

File: src/Symfony/Component/Lock/LockInterface.php (hunk 8 above)
Comment (kevinjhappy) [reply src/Symfony/Component/Lock/LockInterface.php]: Argh I didn't see that thanks

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php (hunk 9 above)
Comment (kevinjhappy) [reply src/Symfony/Component/Security/Guard/GuardAuthenticatorHandler.php]: This arguments is commented since Symfony 3.4, and the class passed final since 4.0, this particular evolution was skipped in the newest version, you are right it's need to be made in 5.0, not now

**Comment by nicolas-grekas** [comment]
> Most of these deprecations won't be triggered if the child method does not call the parent method, so they might not be the best place to trigger them (sometimes, we might not have a better way though)

True. We can't do better I think, unless we create a way to declare such added arguments and trigger a deprecation in e.g. `DebugClassLoader` or via a static analyzer. Could be a useful improvement (but not for this PR of course.)

**Comment by kevinjhappy** [comment]
Thanks @nicolas-grekas and @stof, I fixed Changelog and functions according to your feedbacks.
About the providerKey in `GuardAuthenticatorHandler::authenticateWithToken()`, I started reverting my change to finally uncomment the argument according to Nicolas's last message.

File: UPGRADE-4.2.md [hunk 10]
```diff
@@ -75,6 +75,11 @@ Process
    // alternatively, when a shell wrapper is required
    $process = Process::fromShellCommandline('ls -l');
    ```
+ * In `Bridge/Monolog` the methods `DebugProcessor::getLogs()`, `DebugProcessor::countErrors()`, `Logger::getLogs()` and `Logger::countErrors()` will have one `$request` argument in version 5.0, not defining it is deprecated since Symfony 4.2.
+ * In `Component/BrowserKit`, the `Client::submit()` method will have a new `$serverParameters` argument in version 5.0, not defining it is deprecated since version 4.2.
+ * In `Component/DomCrawler`, the `Crawler::children()` method will have one `$selector` argument in version 5.0, not defining it is deprecated since version 4.2.
+ * In `Component/Finder`, the `Finder::sortByName()` method will have one `$useNaturalSort` argument in version 5.0, not defining it is deprecated since version 4.2.
+ * In `Component/Serializer`, the `AbstractNormalizer::handleCircularReference()` method will have two more arguments : `$format` and `$context` in version 5.0, not defining it is deprecated since version 4.2.
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: right now, all these lines are in the "Process" component section. They should be split each in their component's section instead. (same in UPGRADE-5.0.md)

**Comment by GuilhemN** [comment]
I think we can avoid the limitation @stof described using `@param` annotations. I opened https://github.com/symfony/symfony/pull/28329 to suggest this alternative.

File: src/Symfony/Component/BrowserKit/CHANGELOG.md [hunk 11]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+4.2.0
+-----
+
+ * The method `Client::submit()`  will have one `$serverParameters` argument in version 5.0, not definig it
```
Comment (GuilhemN) [inline src/Symfony/Component/BrowserKit/CHANGELOG.md]: `defining`

Comment (nicolas-grekas) [reply src/Symfony/Component/BrowserKit/CHANGELOG.md]: (and double space :) )

Comment (kevinjhappy) [reply src/Symfony/Component/BrowserKit/CHANGELOG.md]: missed it, done :)

File: UPGRADE-4.2.md [hunk 12]
```diff
@@ -149,3 +169,4 @@ Serializer
 
  * Relying on the default value (false) of the "as_collection" option is deprecated since 4.2.
    You should set it to false explicitly instead as true will be the default value in 5.0.
+ * The `AbstractNormalizer::handleCircularReference()` method will have two more arguments : `$format` and `$context` in version 5.0, not defining it is deprecated since version 4.2.
```
Comment (nicolas-grekas) [inline UPGRADE-4.2.md]: not defining **them**

File: UPGRADE-5.0.md [hunk 13]
```diff
@@ -141,6 +161,11 @@ SecurityBundle
  * The `security.authentication.trust_resolver.anonymous_class` parameter has been removed.
  * The `security.authentication.trust_resolver.rememberme_class` parameter has been removed.
 
+Serializer
+----------
+
+ * The `AbstractNormalizer::handleCircularReference()` method will have two more arguments : `$format` and `$context` in version 5.0, not defining it is deprecated since version 4.2.
```
Comment (nicolas-grekas) [inline UPGRADE-5.0.md]: not defining them

File: src/Symfony/Bridge/Monolog/CHANGELOG.md [hunk 14]
```diff
@@ -5,6 +5,8 @@ CHANGELOG
 -----
 
  * added `ProcessorInterface`: an optional interface to allow autoconfiguration of Monolog processors
+ * The methods `DebugProcessor::getLogs()`, `DebugProcessor::countErrors()`, `Logger::getLogs()` and `Logger::countErrors()` 
+ will have one `$request` argument in version 5.0, not defining it is deprecated since Symfony 4.2.
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/CHANGELOG.md]: missing indentation

File: src/Symfony/Component/BrowserKit/CHANGELOG.md [hunk 15]
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+4.2.0
+-----
+
+ * The method `Client::submit()` will have one `$serverParameters` argument 
+ in version 5.0, not defining it is deprecated since version 4.2
```
Comment (nicolas-grekas) [inline src/Symfony/Component/BrowserKit/CHANGELOG.md]: bad indentation

File: src/Symfony/Component/Security/Guard/Tests/GuardAuthenticatorHandlerTest.php [hunk 16]
```diff
@@ -146,14 +146,15 @@ public function testSessionStrategyIsCalled()
 
     public function testSessionStrategyIsNotCalledWhenStateless()
     {
+        $providerKey = 'some_provider_key';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Guard/Tests/GuardAuthenticatorHandlerTest.php]: can be reverted, isn't it? (same below in the same file)

File: src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php [hunk 17]
```diff
@@ -64,7 +64,7 @@ public function testHandleSuccess()
         $this->guardAuthenticatorHandler
             ->expects($this->once())
             ->method('authenticateWithToken')
-            ->with($authenticateToken, $this->request);
+            ->with($authenticateToken, $this->request, $providerKey);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php]: can be reverted, isn't it?

File: src/Symfony/Component/Serializer/CHANGELOG.md [hunk 18]
```diff
@@ -8,6 +8,8 @@ CHANGELOG
 * added support for XML comment encoding (encoding `['#comment' => ' foo ']` results `<!-- foo -->`)
 * added optional `int[] $encoderIgnoredNodeTypes` argument to `XmlEncoder::__construct` to configure node types to be
   ignored during encoding.
+* the `AbstractNormalizer::handleCircularReference()` method will have two more arguments : `$format` and 
+  `$context` in version 5.0, not defining it is deprecated since version 4.2.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Serializer/CHANGELOG.md]: not defining them

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
LGTM, here are some remaining comments.

File: src/Symfony/Component/Serializer/CHANGELOG.md (hunk 18 above)
Comment (kevinjhappy) [reply src/Symfony/Component/Serializer/CHANGELOG.md]: thanks @nicolas-grekas fixed the typo and revert unit tests :)

File: UPGRADE-5.0.md (hunk 13 above)
Comment (nicolas-grekas) [reply UPGRADE-5.0.md]: ping ;) (rebase needed also + squash if you want)

**Review by nicolas-grekas — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @kevinjhappy.
